rfcs/2021-11-12-9811-vrl-vm.md

```diff
@@ -0,0 +1,347 @@
+
+- The code is a bit more complex. With an AST that is walked it is fairly
+  apparent what the code will be doing at any point. With a Vm, this is not the
+  case, it is harder to look at the instructions in the Vm and follow back to
+  what part of the VRL code is being evaluated. We will need to write some
+  extensive debugging tools to allow for decent introspection into the Vm.
+
+  However, VRL is a very simple language, which does also mean the VM will be
+  simple.
+
+- We lose some safety that we get from the Rust compiler. There will need
+  to be significant fuzz testing to ensure that the code runs correctly under
+  all circumstances.
+
+  It should be noted however that the entire VM will not require any _unsafe_
+  Rust code.
+
+- Currently each stdlib function is responsible for evaluating their own
+  parameters. This allows parameters to be lazily evaluated. With the Vm, the
+  parameters will need to be evaluated up front and the stack passed into the
+  function. This could impact performance.
+
+## Prior Art
+
+- GoScript - An implementation of Go using a bytecode Vm,
+  https://github.com/oxfeeefeee/goscript
+
+- CPython - https://github.com/python/cpython
+
+## Alternatives
+
+### WASM
+
+Instead of writing our own VM we could transpile VRL code to WASM. WASM is a
+mature and performant VM that could give us near native speeds.
+
+The issues with using WASM are that it would require data to be serialized in
+order to move between Vector and WASM. This would incur a significant
+performance penalty.
+
+WASM provides a lot more functionality and complexity than is required for VRL.
+By developing our own VM we will be able to customize it purely for the usecase
+required by VRL. This allows us to implement very VRL specific OpCodes which
+should allow the bytecode stay simple.
+
+### Represent additional data in the OpCode
+
+The current suggestion represents each OpCode as a single instruction in the
+instruction list. An OpCode with it's data occupies two instructions.
+
+For example, with the current suggestion to load a constant three things have
+to happen:
+
+1. Add the Constant OpCode to the instructions.
+2. Add the actual constant value to the list of constants.
+3. Add the index of the constant in the constant list to the list of constants.
+
+#### Include value in the OpCode
+
+These three things could be represented as a single instruction if we included
+the constant value in the OpCode:
+
+```rust
+enum OpCode {
+    ...
+    Constant(Box<Value>)
+    ...
+}
+```
+
+This will increase `size_of::<OpCode>` from 8 to 16, potentially doubling the
+memory size of the instruction list.
+
+It also means we do not need to store a separate list of constants. On the plus
+side, this is one less lookup at runtime to load the constant. On the down side,
+we wouldn't be able to dedupe the constant, so for example, if the code uses the
+same string twice, it would be represented twice in the bytecode.
```

Comment:
Worse, since the values are boxed, you have now introduced a non-linear indirection to get at the value, which is exactly the kind of memory accesses this RFC is trying to avoid.